### PR TITLE
Exceptions: consistency, unit tests, docs, examples

### PR DESCRIPTION
**Pull request by thomascorthals**
As discussed in #747, there were some inconsistencies in the way Solarium threw exceptions.

## TL;DR

- I've added `LogicExceptionInterface` and `RuntimeExceptionInterface`
- and made sure all thrown exceptions are "ours"
- while maintaining full backward compatibility,
- added examples and docs on catching them,
- and provided unit tests to prove our code and docs line up.

## What I didn't do

I didn't change the inheritance tree of the exceptions.
- Building our own tree under `Solarium\Exception` would break BC for users who are e.g. catching `Solarium\Exception\OutOfBoundsException` as a base `OutOfBoundsException`.
- Having `Solarium\Exception\StreamException` extend `Solarium\Exception\RuntimeException` would break BC for users who are catching it as a base `UnexpectedValueException`.
- Having `Solarium\Exception\HttpException` extend `Solarium\Exception\RuntimeException` wouldn't break BC, but doesn't seem terribly consistent at this point.

## What I did do

I've added `LogicExceptionInterface` and `RuntimeExceptionInterface` marker interfaces that extend `ExceptionInterface`. This adds a level of granularity in distinguishing exceptions while maintaining BC.

I've replaced the handful of base exceptions that were thrown with "proper" Solarium exceptions. They extend their base counterpart so this doesn't break BC.

I've added `use` statements for all exceptions in `src` and `tests` instead of scattering fully qualified names throughout the code.

I've added an example that demonstrates how you can distinguish Solarium exceptions from those thrown by other libraries. I've also written short documentation, emphasising that you must at least catch an `HttpException` for every request.

Examples that used `catch (Solarium\Exception $e)` have been updated to `catch (Exception $e)`. They ran fine as long as there is no exception, but would result in a fatal error when there is because they didn't actually match the exception we are trying to catch.

Since I've "formalised" the way to handle our exceptions in the docs, I've written unit tests that prove the inheritance model is what we describe. This safeguards us against introducing backward incompatible changes to the inheritance tree.

Test coverage for `HttpException` and `StreamException` has been rounded out to cover all method signatures.

### CODE REVIEW

**Comment by mkalkbrenner** [comment]
I thought twice about this patch.
Since it maintains BC and increases the quality I'm of the opinion that we should include it in the final solarium 6.0.0 release even if we already reached rc1.
